From c21607c3d0b473a3f8663f8f08e0ace7aac9298c Mon Sep 17 00:00:00 2001
From: Marcelo Tosatti <mtosatti@redhat.com>
Date: Mon, 8 Mar 2010 20:22:08 -0300
Subject: [PATCH] Fix PkgLength calculation for the SSDT

RH-Author: Marcelo Tosatti <mtosatti@redhat.com>
Message-id: <20100308202208.GB14376@amt.cnet>
Patchwork-id: 7598
O-Subject: [RHEL6 seabios] Fix PkgLength calculation for the SSDT
Bugzilla: 571553
RH-Acked-by: Gleb Natapov <gleb@redhat.com>
RH-Acked-by: Michael S. Tsirkin <mst@redhat.com>
RH-Acked-by: Justin M. Forbes <jforbes@redhat.com>

commit 9fb3f4d950744e97cc655b7d7b523d8bf101e4a0
Author: Magnus Christensson <mch@virtutech.com>
Date:   Wed Nov 25 16:26:58 2009 +0100

Fix PkgLength calculation for the SSDT.

Signed-off-by: Magnus Christensson <mch@virtutech.com>

BZ: 571553

Signed-off-by: Eduardo Habkost <ehabkost@redhat.com>
---
 src/acpi.c |    6 ++++--
 1 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/acpi.c b/src/acpi.c
index f613b03..244536a 100644
--- a/src/acpi.c
+++ b/src/acpi.c
@@ -429,10 +429,12 @@ build_ssdt(void)
     // build processor scope header
     *(ssdt_ptr++) = 0x10; // ScopeOp
     if (cpu_length <= 0x3e) {
+        /* Handle 1-4 CPUs with one byte encoding */
         *(ssdt_ptr++) = cpu_length + 1;
     } else {
-        *(ssdt_ptr++) = 0x7F;
-        *(ssdt_ptr++) = (cpu_length + 2) >> 6;
+        /* Handle 5-314 CPUs with two byte encoding */
+        *(ssdt_ptr++) = 0x40 | ((cpu_length + 2) & 0xf);
+        *(ssdt_ptr++) = (cpu_length + 2) >> 4;
     }
     *(ssdt_ptr++) = '_'; // Name
     *(ssdt_ptr++) = 'P';
-- 
1.6.3.rc4.29.g8146

